hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -159,6 +159,24 @@ func (r *NodePoolReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 				return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from nodepool: %w", err)
 			}
 		}
+
+		//Ensure all machines are gone before deleting related machine secrets
```

Comment:
this logic is already existing in `r.delete()` see `listSecrets` and `deleteMachineSet()`, please handle order in `r.delete()` if needed.

Reply:
I could be missing it, but I don't think r.delete() is handling this. The issue seemed to be that the userSecret needs to stay around until all of the machine sets in the nodepool being deleted are truly gone. I thought the logic in r.delete() was just deleting a single machine set.

I can move this logic to r.delete() though.

Reply:
I was missing how deleteMachineSet() worked but I understand better now. 

I have the deleteSecrets commented out for now because if the secrets are deleted prior to the machines being fully deleted, it causes this error in the capi controller, `failed to retrieve bootstrap data secret for AWSMachine clusters-brcox-hypershift-x86/brcox-hypershift-x86-work-1-glht8: Secret \"user-data-brcox-hypershift-x86-work-1-12f2ebfe\" not found" `. Moving `deleteNodePoolSecrets` to the end of `delete` does not solve the issue either.

Reply:
In additional to what is implemented in the PR currently, I also have this working locally at the `nodepool_controller.go` level:
```
if nodePool.Spec.Platform.AWS != nil {
	allMachinesTerminated := false
	allMachinesTerminated, err := r.allMachinesTerminated(nodePool)
	if err != nil {
		return ctrl.Result{}, err
	}

	if !allMachinesTerminated {
		return ctrl.Result{Requeue: true}, nil
	}

	r.deleteNodePoolSecrets(ctx, nodePool)
}

// Now we can remove the finalizer.
if controllerutil.ContainsFinalizer(nodePool, finalizer) {
	controllerutil.RemoveFinalizer(nodePool, finalizer)
	if err := r.Update(ctx, nodePool); err != nil {
		return ctrl.Result{}, fmt.Errorf("failed to remove finalizer from nodepool: %w", err)
	}
}
log.Info("Deleted nodepool", "name", req.NamespacedName)
```

I'm open to suggestions or other paths, but basically it seems like we can't delete the NodePool secrets until all the NodePool machines are gone.